org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/FinallyFilter.java

```diff
@@ -0,0 +1,143 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2017 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.internal.analysis.filter;
+
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.tree.AbstractInsnNode;
+import org.objectweb.asm.tree.JumpInsnNode;
+import org.objectweb.asm.tree.MethodNode;
+import org.objectweb.asm.tree.TryCatchBlockNode;
+import org.objectweb.asm.tree.VarInsnNode;
+
+/**
+ * Filters duplicates of finally blocks that compiler generates.
+ */
+public final class FinallyFilter implements IFilter {
+
+	public void filter(final String className, final String superClassName,
+			final MethodNode methodNode, final IFilterOutput output) {
+		for (TryCatchBlockNode tryCatchBlock : methodNode.tryCatchBlocks) {
+			if (tryCatchBlock.type == null) {
+				filter(tryCatchBlock, output);
+				filterEcj(tryCatchBlock, output);
+			}
+		}
+	}
+
+	private void filter(final TryCatchBlockNode tryCatchBlock,
+			final IFilterOutput output) {
+		final AbstractInsnNode n = next(tryCatchBlock.end);
+		final AbstractInsnNode e = next(tryCatchBlock.handler);
+
+		final int size = size(e);
+		if (size == 0) {
+			return;
+		}
+		if (!isSame(size, next(e), n)) {
+			return;
+		}
+		filter(output, size, e, n);
+	}
+
+	private static void filter(final IFilterOutput output, final int size,
+			AbstractInsnNode e, AbstractInsnNode n) {
+		output.ignore(e, e);
+		e = next(e);
+		for (int i = 0; i < size; i++) {
+			output.merge(e, n);
+			n = next(n);
+			e = next(e);
+		}
+		output.ignore(e, next(e));
+		if (n != null && n.getOpcode() == Opcodes.GOTO) {
```

Comment:
:disappointed: however in case of ECJ this example demonstrates that we miss merge of instructions that are at the exit of a loop without `break`. And for slightly modified example same happens with javac, but not with ECJ:

```
void fun() {
  while (t())
    try {
      if (t())
	break;
      nop();
    } finally {
      nop();
    }
}
```